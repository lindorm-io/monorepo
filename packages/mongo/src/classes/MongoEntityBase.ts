import { IMongoEntity } from "../interfaces";

export class MongoEntityBase implements IMongoEntity {
  public readonly id!: string;
  public readonly rev!: number;
  public readonly seq!: number;
  public readonly createdAt!: Date;
  public readonly deletedAt!: Date | null;
  public readonly updatedAt!: Date;

  public expiresAt!: Date | null;
}
