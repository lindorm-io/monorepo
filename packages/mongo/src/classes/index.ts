export * from "./MongoBucket";
export * from "./MongoFileBase";
export * from "./MongoRepository";
export * from "./MongoSource";
