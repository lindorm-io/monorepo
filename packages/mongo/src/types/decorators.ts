export type FileDecoratorOptions = {
  chunkSizeBytes?: number;
  database?: string;
  name?: string;
  namespace?: string;
};
