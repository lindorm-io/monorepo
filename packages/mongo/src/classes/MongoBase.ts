import { ILogger } from "@lindorm/logger";
import { Collection, Db, Document, MongoClient } from "mongodb";
import { MongoRepositoryError } from "../errors";
import { MongoBaseIndex, MongoBaseOptions } from "../types";

export abstract class MongoBase<D extends Document> {
  protected readonly client: MongoClient;
  protected readonly collectionName: string;
  protected readonly databaseName: string;
  protected readonly indexes: Array<MongoBaseIndex>;
  protected readonly logger: ILogger;

  protected _collection: Collection<D> | undefined;
  protected _database: Db | undefined;

  protected constructor(options: MongoBaseOptions) {
    this.logger = options.logger.child(["MongoBase"]);

    this.collectionName = options.collection;
    this.databaseName = options.database;
    this.indexes = options.indexes;
    this.client = options.client;
  }

  // getters

  protected get collection(): Collection<any> {
    if (!this._collection) {
      this._collection = this.database.collection(this.collectionName);
    }
    return this._collection;
  }

  protected get database(): Db {
    if (!this._database) {
      this._database = this.client.db(this.databaseName);
    }
    return this._database;
  }

  // public

  public async setup(): Promise<void> {
    for (const { index, options } of this.indexes) {
      this.logger.debug("Creating index", {
        collection: this.collectionName,
        index,
        options,
      });

      try {
        const result = await this.collection.createIndex(index, options);

        this.logger.silly("Index created", { result });
      } catch (error: any) {
        this.logger.silly("Mongo setup error", error);

        throw new MongoRepositoryError(error.message, {
          debug: { collection: this.collectionName, index, options },
          error,
        });
      }
    }
  }
}
