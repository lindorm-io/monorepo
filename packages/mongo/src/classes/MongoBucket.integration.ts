import { createMockLogger } from "@lindorm/logger";
import { randomUUID } from "crypto";
import { createReadStream } from "fs";
import MockDate from "mockdate";
import { MongoClient } from "mongodb";
import { join } from "path";
import { Readable } from "stream";
import { TestBucket } from "../__fixtures__/test-bucket";
import { TestFile } from "../__fixtures__/test-file";
import { getReadStreamContent } from "../__fixtures__/utils";
import { MongoBucketError } from "../errors";
import { FileUpload } from "../types";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate);

describe("MongoBucket", () => {
  let client: MongoClient;
  let bucket: TestBucket;
  let stream: () => Readable;
  let metadata: FileUpload<TestFile>;

  beforeAll(async () => {
    client = new MongoClient("mongodb://root:example@localhost/admin?authSource=admin");
    await client.connect();
    bucket = new TestBucket(client, createMockLogger());
  });

  beforeEach(async () => {
    stream = () =>
      createReadStream(join(__dirname, "..", "__fixtures__", "test-file-upload.txt"));

    metadata = {
      encoding: "7bit",
      extraOne: 1,
      extraTwo: 2,
      hash: "hash",
      hashAlgorithm: "sha256",
      mimeType: "text/plain",
      name: randomUUID(),
      originalName: "test-file-upload.txt",
      size: 18,
      strategy: "local",
    };
  });

  afterAll(async () => {
    await client.close();
  });

  test("should setup", async () => {
    await expect(bucket.setup()).resolves.not.toThrow();
  }, 30000);

  test("should delete many files", async () => {
    const f1 = await bucket.upload(stream(), metadata);
    const f2 = await bucket.upload(stream(), metadata);

    await expect(bucket.delete({ name: f1.name })).resolves.not.toThrow();

    await expect(bucket.findOne({ filename: f1.filename })).resolves.toBeNull();
    await expect(bucket.download(f1.filename)).rejects.toThrow();

    await expect(bucket.findOne({ filename: f2.filename })).resolves.toBeNull();
    await expect(bucket.download(f2.filename)).rejects.toThrow();
  });

  test("should download file by name", async () => {
    const result = await bucket.upload(stream(), metadata);
    const download = await bucket.download(result.filename);

    expect(download.chunkSize).toEqual(261120);
    expect(download.filename).toEqual(result.filename);
    expect(download.length).toEqual(18);
    expect(download.mimeType).toEqual("text/plain");
    expect(download.originalName).toEqual("test-file-upload.txt");
    expect(download.uploadDate).toEqual(MockedDate);

    await expect(getReadStreamContent(download.stream)).resolves.toEqual(
      "test file content\n",
    );
  });

  test("should find many files", async () => {
    const f1 = await bucket.upload(stream(), metadata);
    const f2 = await bucket.upload(stream(), metadata);

    await expect(bucket.find({ name: f1.name })).resolves.toEqual([f1, f2]);
    await expect(bucket.find({ name: randomUUID() })).resolves.toEqual([]);
  });

  test("should find one file", async () => {
    const file = await bucket.upload(stream(), metadata);

    await expect(bucket.findOne({ name: file.name })).resolves.toEqual(file);
    await expect(bucket.findOne({ name: randomUUID() })).resolves.toBeNull();
  });

  test("should find one or fail", async () => {
    const file = await bucket.upload(stream(), metadata);

    await expect(bucket.findOneOrFail({ name: file.name })).resolves.toEqual(file);
    await expect(bucket.findOneOrFail({ name: randomUUID() })).rejects.toThrow(
      MongoBucketError,
    );
  });

  test("should find one file by filename", async () => {
    const file = await bucket.upload(stream(), metadata);

    await expect(bucket.findOne({ filename: file.filename })).resolves.toEqual(file);
    await expect(bucket.findOne({ filename: randomUUID() })).resolves.toBeNull();
  });

  test("should find one file by filename or fail", async () => {
    const file = await bucket.upload(stream(), metadata);

    await expect(bucket.findOneOrFail({ filename: file.filename })).resolves.toEqual(
      file,
    );
    await expect(bucket.findOneOrFail({ filename: randomUUID() })).rejects.toThrow(
      MongoBucketError,
    );
  });

  test("should upload file", async () => {
    await expect(bucket.upload(stream(), metadata)).resolves.toEqual({
      chunkSize: 261120,
      encoding: "7bit",
      extraOne: 1,
      extraTwo: 2,
      filename: expect.any(String),
      hash: "hash",
      hashAlgorithm: "sha256",
      length: 18,
      mimeType: "text/plain",
      name: metadata.name,
      originalName: "test-file-upload.txt",
      size: 18,
      strategy: "local",
      uploadDate: MockedDate,
    });
  });
});
