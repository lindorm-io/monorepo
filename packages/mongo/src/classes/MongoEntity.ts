import { EntityBase } from "@lindorm/entity";
import { IMongoEntity } from "../interfaces";
import { MongoEntityConfig } from "../types";

export class MongoEntity extends EntityBase implements IMongoEntity {
  public readonly rev!: number;
  public readonly seq!: number;
  public readonly deletedAt!: Date | null;
  public expiresAt!: Date | null;
}

export const MONGO_ENTITY_CONFIG: MongoEntityConfig<MongoEntity> = {
  revisionAttribute: "rev",
  sequenceAttribute: "seq",
  deleteAttribute: "deletedAt",
  ttlAttribute: "expiresAt",
};
