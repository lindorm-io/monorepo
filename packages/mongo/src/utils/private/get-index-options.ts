import { EntityMetadata } from "@lindorm/entity";
import { CreateIndexesOptions } from "mongodb";
import { MongoBaseIndex } from "../../types";

const getPartialFilterExpression = (
  metadata: EntityMetadata,
  keys: Array<string>,
): Partial<CreateIndexesOptions> => {
  const nullable = metadata.columns
    .filter((a) => a.nullable && keys.includes(a.key))
    .map((a) => a.key);

  const finite = metadata.columns
    .filter((a) => a.min && a.min > 0 && keys.includes(a.key))
    .map((a) => a.key);

  if (!nullable.length && !finite.length) {
    return {};
  }

  return {
    partialFilterExpression: {
      ...nullable.reduce((a, k) => ({ ...a, [k]: { $exists: true } }), {}),
      ...finite.reduce((a, k) => ({ ...a, [k]: { $gt: 0 } }), {}),
    },
  };
};

export const getIndexOptions = (metadata: EntityMetadata): Array<MongoBaseIndex> => {
  const result: Array<MongoBaseIndex> = [];

  result.push({
    index: metadata.primaryKeys.reduce((a, k) => ({ ...a, [k]: 1 }), {}),
    options: {
      unique: true,
      ...getPartialFilterExpression(metadata, metadata.primaryKeys),
    },
  });

  for (const index of metadata.indexes) {
    result.push({
      index: Object.entries(index.index).reduce(
        (acc, [key, direction]) => ({
          ...acc,
          [key]: direction === "asc" ? 1 : direction === "desc" ? -1 : direction,
        }),
        {},
      ),
      options: {
        ...(index.name ? { name: index.name } : {}),
        ...getPartialFilterExpression(metadata, Object.keys(index.index)),
        ...index.options,
      },
    });
  }

  for (const unique of metadata.uniques) {
    result.push({
      index: unique.keys.reduce((acc, key) => ({ ...acc, [key]: 1 }), {}),
      options: {
        ...(unique.name ? { name: unique.name } : {}),
        unique: true,
        ...getPartialFilterExpression(metadata, unique.keys),
      },
    });
  }

  return result;
};
