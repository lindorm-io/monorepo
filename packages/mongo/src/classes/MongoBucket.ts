import {
  defaultCreateEntity,
  defaultValidateEntity,
  getCollectionName,
  globalEntityMetadata,
} from "@lindorm/entity";
import { ILogger } from "@lindorm/logger";
import { Constructor, DeepPartial, Dict } from "@lindorm/types";
import { randomUUID } from "crypto";
import { DeleteOptions, Filter, FindOptions, GridFSBucket } from "mongodb";
import { Readable } from "stream";
import { FileExtra } from "../decorators";
import { MongoBucketError } from "../errors";
import { IMongoBucket, IMongoFile } from "../interfaces";
import { GridFSDocument } from "../interfaces/private";
import { FileDownload, FileUpload, MongoBucketOptions } from "../types";
import {
  getBucketChunkSize,
  getBucketCollectionName,
  getBucketMetadataIndexes,
  getIndexOptions,
} from "../utils/private";
import { MongoBase } from "./MongoBase";

export class MongoBucket<F extends IMongoFile>
  extends MongoBase<F>
  implements IMongoBucket<F>
{
  private readonly FileConstructor: Constructor<F>;
  private readonly bucketName: string;
  private readonly chunkSizeBytes: number | null;
  private _grid: GridFSBucket | undefined;

  protected readonly logger: ILogger;

  public constructor(options: MongoBucketOptions<F>) {
    const metadata = globalEntityMetadata.get<FileExtra>(options.File);
    const database = metadata.entity.database || options.database;

    if (!database) {
      throw new MongoBucketError("Database name not found", {
        debug: { metadata, options },
      });
    }

    super({
      client: options.client,
      collection: getBucketCollectionName(options.File, options),
      database,
      logger: options.logger,
      indexes: getIndexOptions({
        ...metadata,
        indexes: getBucketMetadataIndexes(metadata),
      }),
    });

    this.logger = options.logger.child(["MongoMongoBucket", options.File.name]);

    this.bucketName = getCollectionName(options.File, options);
    this.chunkSizeBytes = getBucketChunkSize(options, metadata);

    this.FileConstructor = options.File;
  }

  // getters

  protected get grid(): GridFSBucket {
    if (!this._grid) {
      this._grid = new GridFSBucket(this.client.db(this.databaseName), {
        bucketName: this.bucketName,
        ...(this.chunkSizeBytes ? { chunkSizeBytes: this.chunkSizeBytes } : {}),
      });
    }

    return this._grid;
  }

  // public

  public async delete(criteria: Filter<F>, options?: DeleteOptions): Promise<void> {
    const start = Date.now();

    try {
      const filter = this.parseFilter(criteria);
      const result = await this.collection.find(filter).toArray();

      for (const document of result) {
        await this.grid.delete(document._id);
      }

      await this.collection.deleteMany(filter, options);

      this.logger.debug("Bucket done: delete", {
        input: {
          filter,
        },
        result: {
          ...result,
          time: Date.now() - start,
        },
      });
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Unable to delete files", { error });
    }
  }

  public async download(filename: string): Promise<FileDownload<F>> {
    try {
      const file = await this.findOneOrFail({ filename } as any);
      const stream = this.grid.openDownloadStreamByName(filename);

      return { ...file, stream };
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Error downloading file", {
        data: { filename },
        error,
      });
    }
  }

  public async find(criteria?: Filter<F>, options?: FindOptions<F>): Promise<Array<F>> {
    const start = Date.now();

    try {
      const filter = this.parseFilter(criteria);
      const result = await this.collection.find(filter, options).toArray();

      this.logger.debug("Bucket done: find", {
        input: {
          criteria,
          filter,
          options,
        },
        result: {
          count: result.length,
          time: Date.now() - start,
        },
      });

      return result.map((document) => this.create(document));
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Unable to find files", { error });
    }
  }

  public async findOne(criteria: Filter<F>, options?: FindOptions<F>): Promise<F | null> {
    const start = Date.now();

    try {
      const filter = this.parseFilter(criteria);
      const document = await this.collection.findOne(filter, options);

      this.logger.debug("Bucket done: findOne", {
        input: {
          criteria,
          filter,
          options,
        },
        result: {
          document,
          time: Date.now() - start,
        },
      });

      if (!document) return null;

      return this.create(document);
    } catch (error: any) {
      this.logger.error("MongoBucket error", error);
      throw new MongoBucketError("Unable to find file", { error });
    }
  }

  public async findOneOrFail(criteria: Filter<F>, options?: FindOptions<F>): Promise<F> {
    const document = await this.findOne(criteria, options);

    if (!document) {
      throw new MongoBucketError("File not found", { debug: { criteria, options } });
    }

    return document;
  }

  public async upload(stream: Readable, metadata: FileUpload<F>): Promise<F> {
    this.logger.debug("Uploading file", { metadata });

    try {
      const filename = randomUUID();
      const upload = this.grid.openUploadStream(filename, { metadata });

      stream.pipe(upload);

      this.logger.debug("File upload stream initialised", { filename });

      await new Promise<void>((resolve, reject) => {
        upload.once("finish", resolve);
        upload.once("error", reject);
      });

      const { done, state, gridFSFile } = upload;

      if (!done) {
        throw new Error("File upload not done");
      }

      if (!gridFSFile) {
        throw new Error("GridFSFile not initialized");
      }

      const { length, chunkSize, uploadDate } = gridFSFile;

      this.logger.error("Bucket done: upload", {
        chunkSize,
        done,
        length,
        metadata,
        state,
        uploadDate,
      });

      const file = await this.findOneOrFail({ filename } as any);

      try {
        this.validate(file);
      } catch (error: any) {
        this.logger.error("Error validating file", error);

        await this.grid.delete(upload.id);
        await this.collection.deleteOne({ _id: upload.id });

        throw new MongoBucketError("Error validating file", { error });
      }

      return file;
    } catch (error: any) {
      this.logger.error("Error uploading file", error);

      throw new MongoBucketError("Error uploading file", { error });
    }
  }

  // private

  private create(document: GridFSDocument): F {
    const { chunkSize, filename, length, uploadDate, metadata = {} } = document;

    const file = defaultCreateEntity<F>(this.FileConstructor, {
      chunkSize,
      filename,
      length,
      uploadDate,
      ...metadata,
    } as DeepPartial<F>);

    this.logger.debug("Created file", { file });

    return file;
  }

  private validate(file: F): void {
    defaultValidateEntity(this.FileConstructor, file);

    this.logger.debug("File validated", { file });
  }

  private parseFilter(criteria: Filter<any> = {}): Filter<any> {
    const parsed: Dict = {};

    for (const [key, value] of Object.entries(criteria)) {
      if (
        key === "chunkSize" ||
        key === "filename" ||
        key === "length" ||
        key === "uploadDate"
      ) {
        parsed[key] = value;
        continue;
      }

      parsed[`metadata.${key}`] = value;
    }

    this.logger.debug("Parsed criteria", { criteria, parsed });

    return parsed;
  }
}
