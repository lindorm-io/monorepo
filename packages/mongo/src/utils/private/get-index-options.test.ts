import { Entity, EntityBase, globalEntityMetadata } from "@lindorm/entity";
import { MongoFileBase } from "../../classes";
import { File } from "../../decorators";
import { getIndexOptions } from "./get-index-options";

describe("getIndexOptions", () => {
  test("should return options for EntityBase", () => {
    @Entity()
    class TestEntityBase extends EntityBase {}

    const metadata = globalEntityMetadata.get(TestEntityBase);

    expect(getIndexOptions(metadata)).toMatchSnapshot();
  });

  test("should return options for MongoFileBase", () => {
    @File()
    class TestMongoFileBase extends MongoFileBase {}

    const metadata = globalEntityMetadata.get(TestMongoFileBase);

    expect(getIndexOptions(metadata)).toMatchSnapshot();
  });
});
