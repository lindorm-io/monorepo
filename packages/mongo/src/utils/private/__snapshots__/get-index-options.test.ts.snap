// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIndexOptions should return options for EntityBase 1`] = `
[
  {
    "index": {
      "id": 1,
    },
    "options": {
      "unique": true,
    },
  },
]
`;

exports[`getIndexOptions should return options for MongoFileBase 1`] = `
[
  {
    "index": {
      "filename": 1,
    },
    "options": {
      "unique": true,
    },
  },
  {
    "index": {
      "mimeType": 1,
    },
    "options": {
      "partialFilterExpression": {
        "mimeType": {
          "$exists": true,
        },
      },
    },
  },
  {
    "index": {
      "originalName": 1,
    },
    "options": {
      "partialFilterExpression": {
        "originalName": {
          "$exists": true,
        },
      },
    },
  },
]
`;
